hudi-common/src/main/java/org/apache/hudi/common/engine/HoodieReaderContext.java

```diff
@@ -262,30 +265,52 @@ private BiFunction<T, Schema, String> virtualKeyExtractor(String[] recordKeyFiel
   /**
    * Gets the ordering value in particular type.
    *
-   * @param record An option of record.
+   * @param record An engine specific record.
    * @param schema The Avro schema of the record.
-   * @param orderingFieldName name of the ordering field
+   * @param orderingFieldName name of the ordering field, if any
    * @return The ordering value.
    */
-  public Comparable getOrderingValue(T record,
-                                     Schema schema,
-                                     Option<String> orderingFieldName) {
+  public Comparable getOrderingValue(T record, Schema schema, Option<String> orderingFieldName) {
     if (orderingFieldName.isEmpty()) {
       return DEFAULT_ORDERING_VALUE;
     }
 
     Object value = getValue(record, schema, orderingFieldName.get());
-    Comparable finalOrderingVal = value != null ? convertValueToEngineType((Comparable) value) : DEFAULT_ORDERING_VALUE;
-    return finalOrderingVal;
+    return value != null ? convertValueToEngineType((Comparable) value) : DEFAULT_ORDERING_VALUE;
   }
 
   /**
    * Constructs a new {@link HoodieRecord} based on the given buffered record {@link BufferedRecord}.
+   * Safe to assume that the buffered record is not a delete.
    *
    * @param bufferedRecord  The {@link BufferedRecord} object with engine-specific row
    * @return A new instance of {@link HoodieRecord}.
    */
-  public abstract HoodieRecord<T> constructHoodieRecord(BufferedRecord<T> bufferedRecord);
+  protected abstract HoodieRecord<T> constructHoodieDataRecord(BufferedRecord<T> bufferedRecord);
+
+  /**
+   * Constructs a new {@link HoodieRecord} based on the given buffered record {@link BufferedRecord}.
+   *
+   * @param bufferedRecord  The {@link BufferedRecord} object with engine-specific row
+   * @return A new instance of {@link HoodieRecord}.
+   */
+  public HoodieRecord<T> constructHoodieRecord(BufferedRecord<T> bufferedRecord) {
+    if (bufferedRecord.isDelete()) {
+      return new HoodieEmptyRecord<>(
```

Comment:
It's dangerous to do this because many mergers just returns empty if the payload data is null, then the event time merging semantics would be lost and we also lost the payload data that been stored in the `BufferedRecord`.

Reply:
This is how the code is currently written, I am just doing some refactoring. 
See these references: [Spark](https://github.com/apache/hudi/blob/master/hudi-client/hudi-spark-client/src/main/scala/org/apache/hudi/BaseSparkInternalRowReaderContext.java#L96) and [Flink](https://github.com/apache/hudi/blob/master/hudi-flink-datasource/hudi-flink/src/main/java/org/apache/hudi/table/format/FlinkRowDataReaderContext.java#L151).

Are you saying that this is already wrong and needs to be fixed? If so, is the solution check if the data is present instead of simply whether it is a delete so we carry through as much context as possible?

Reply:
I've updated this to check for `null` instead of `isDelete`